In [49]:
from constants import read_csv

In [50]:
appareances = read_csv("appearances.csv")
players = read_csv("players.csv")

In [51]:
appareances.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1507351 entries, 0 to 1507350
Data columns (total 13 columns):
 #   Column                  Non-Null Count    Dtype 
---  ------                  --------------    ----- 
 0   appearance_id           1507351 non-null  object
 1   game_id                 1507351 non-null  int64 
 2   player_id               1507351 non-null  int64 
 3   player_club_id          1507351 non-null  int64 
 4   player_current_club_id  1507351 non-null  int64 
 5   date                    1507351 non-null  object
 6   player_name             1507027 non-null  object
 7   competition_id          1507351 non-null  object
 8   yellow_cards            1507351 non-null  int64 
 9   red_cards               1507351 non-null  int64 
 10  goals                   1507351 non-null  int64 
 11  assists                 1507351 non-null  int64 
 12  minutes_played          1507351 non-null  int64 
dtypes: int64(9), object(4)
memory usage: 149.5+ MB


In [52]:
appareances.head()

appearance_id  game_id  player_id  player_club_id  player_current_club_id  \
0   2231978_38004  2231978      38004             853                     235   
1   2233748_79232  2233748      79232            8841                    2698   
2   2234413_42792  2234413      42792            6251                     465   
3   2234418_73333  2234418      73333            1274                    6646   
4  2234421_122011  2234421     122011             195                    3008   

         date       player_name competition_id  yellow_cards  red_cards  \
0  2012-07-03  Aurélien Joachim            CLQ             0          0   
1  2012-07-05    Ruslan Abyshov            ELQ             0          0   
2  2012-07-05       Sander Puri            ELQ             0          0   
3  2012-07-05   Vegar Hedenstad            ELQ             0          0   
4  2012-07-05  Markus Henriksen            ELQ             0          0   

   goals  assists  minutes_played  
0      2        0              90  
1      0        0              90  
2      0        0              45  
3      0        0              90  
4      0        1              90

# About `player_current_club_id` field...
I cannot understand what's the difference between this field and player_club_id viewing only this dataset, so i decided to join with `players` dataset to confirm my intuition that `player_club_id` refer to the club the player was in when appereance was recorded and `player_current_club_id` refers to the same `club_id` that appear in the `players` dataset


In [53]:
players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30302 entries, 0 to 30301
Data columns (total 23 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   player_id                             30302 non-null  int64  
 1   first_name                            28337 non-null  object 
 2   last_name                             30302 non-null  object 
 3   name                                  30302 non-null  object 
 4   last_season                           30302 non-null  int64  
 5   current_club_id                       30302 non-null  int64  
 6   player_code                           30302 non-null  object 
 7   country_of_birth                      27613 non-null  object 
 8   city_of_birth                         28099 non-null  object 
 9   country_of_citizenship                29759 non-null  object 
 10  date_of_birth                         30255 non-null  object 
 11  sub_position   

In [54]:
join_table = appareances.merge(players, on="player_id" , how="inner")
join_table

appearance_id  game_id  player_id  player_club_id  \
0          2231978_38004  2231978      38004             853   
1          2231980_38004  2231980      38004             853   
2          2244350_38004  2244350      38004             853   
3          2244351_38004  2244351      38004             853   
4          2246629_38004  2246629      38004             853   
...                  ...      ...        ...             ...   
1506990   4222126_555166  4222126     555166             232   
1506991  4094688_1111912  4094688    1111912            1421   
1506992   4098878_580434  4098878     580434            2727   
1506993   4172110_624258  4172110     624258              31   
1506994   4172154_696027  4172154     696027              15   

         player_current_club_id        date       player_name competition_id  \
0                           235  2012-07-03  Aurélien Joachim            CLQ   
1                           235  2012-07-10  Aurélien Joachim            CLQ   
2                           235  2012-07-17  Aurélien Joachim            CLQ   
3                           235  2012-07-24  Aurélien Joachim            CLQ   
4                           235  2012-08-01  Aurélien Joachim            CLQ   
...                         ...         ...               ...            ...   
1506990                     232  2023-11-29     Maksim Laykin            RUP   
1506991                    1421  2023-11-26      Adama Bojang            FR1   
1506992                    2727  2023-11-26       Tobe Leysen            BE1   
1506993                      31  2023-11-30     Conor Bradley             EL   
1506994                      15  2023-11-30      Ayman Aourir             EL   

         yellow_cards  red_cards  ...   foot  height_in_cm  \
0                   0          0  ...   left         185.0   
1                   0          0  ...   left         185.0   
2                   0          0  ...   left         185.0   
3                   0          0  ...   left         185.0   
4                   0          0  ...   left         185.0   
...               ...        ...  ...    ...           ...   
1506990             0          0  ...  right         179.0   
1506991             0          0  ...  right         191.0   
1506992             0          0  ...   left         189.0   
1506993             0          0  ...  right         181.0   
1506994             0          0  ...   both         172.0   

         market_value_in_eur highest_market_value_in_eur  \
0                        NaN                    600000.0   
1                        NaN                    600000.0   
2                        NaN                    600000.0   
3                        NaN                    600000.0   
4                        NaN                    600000.0   
...                      ...                         ...   
1506990             500000.0                    500000.0   
1506991             500000.0                    500000.0   
1506992             300000.0                    400000.0   
1506993            1500000.0                   1500000.0   
1506994                  NaN                         NaN   

        contract_expiration_date           agent_name  \
0                            NaN      ER Sport Agency   
1                            NaN      ER Sport Agency   
2                            NaN      ER Sport Agency   
3                            NaN      ER Sport Agency   
4                            NaN      ER Sport Agency   
...                          ...                  ...   
1506990      2026-06-30 00:00:00                  NaN   
1506991      2028-06-30 00:00:00  SPOCS Global Sports   
1506992      2027-06-30 00:00:00     PROFI-MANAGEMENT   
1506993      2024-06-30 00:00:00                  NaN   
1506994      2025-06-30 00:00:00      Bektas Demirtas   

                                                 image_url  \
0        https://img.a.transfermarkt.technology/portrai...   
1     

In [55]:
query1 = join_table.query('player_club_id == current_club_id')
len(query1) == len(join_table)


False

The `player_club_id` don't coincide to the id of the current player club found in `players` dataset, let's test using 
the `player_current_club_id` attribute

In [56]:
query2 = join_table.query('player_current_club_id == current_club_id')
len(query2) == len(join_table)

False

The `player_current_club_id` don't coincide to the id of the current player club found in `players` dataset. So I must assume that also `player_current_club_id` have nothing to do with `current_club_id`.
Let's print the two lengths:

In [57]:
print(f'query1: {len(query1)}')
print(f'query2: {len(query2)}')


query1: 710018
query2: 1505779


In [58]:
print(f'cases with player_club_id not equal to current_club_id = {len(join_table) - len(query1)}')
print(f'cases with player_current_club_id not equal to current_club_id = {len(join_table) - len(query2)}')

cases with player_club_id not equal to current_club_id = 796977
cases with player_current_club_id not equal to current_club_id = 1216


`player_current_club_id` is "more near" to be equivalent to `current_club_id` (1216 exceptions) than `player_club_id` (796977 exceptions) and there also are cases in wich `current_club_id`, `player_current_club_id`, `player_club_id` are all different

# Conclusion 
I will put `player_current_club_id` equal to  `current_club_id` of the player, for my application it not make any sense to have three different references to some club for the same player

In [59]:
join_table['player_current_club_id'] = join_table['current_club_id']

In [60]:
appareances = join_table[list(appareances.keys())]

# Verifing

In [61]:
res = appareances.merge(players, on="player_id", how="inner")[['player_current_club_id' , 'current_club_id']]
#see if for all rows "player_current_club_id" equals to  'current_club_id'
res['player_current_club_id'].equals(res['current_club_id'])

True

In [62]:
appareances.to_csv('../../datasets/new_appearances.csv', index=False)

# Second check

In [63]:
new_ds = pd.read_csv("../../datasets/new_appearances.csv")
join_tb = new_ds.merge(players, on="player_id", how="inner")[['player_current_club_id' , 'current_club_id']]
join_tb['player_current_club_id'].equals(join_tb['current_club_id'])


True